File: macos/Sources/AeroPointAgent/Protocol/MessageValidator.swift
import Foundation

public final class MessageValidator {
    private var seenSequences: Set<Int> = []
    private let decoder = JSONDecoder()

    public init() {}

    public func validate(_ data: Data) throws -> AeroPointMessage {
        let raw: RawMessage
        do {
            raw = try decoder.decode(RawMessage.self, from: data)
        } catch DecodingError.keyNotFound(let key, _) {
            throw MessageValidationError.missingField(key.stringValue)
        } catch {
            throw MessageValidationError.invalidJSON
        }

        guard let seq = raw.seq else {
            throw MessageValidationError.missingField("seq")
        }
        guard seenSequences.insert(seq).inserted else {
            throw MessageValidationError.duplicateSequence(seq)
        }

        switch raw.type {
        case "mouse.move":
            guard let dx = raw.dx else { throw MessageValidationError.missingField("dx") }
            guard let dy = raw.dy else { throw MessageValidationError.missingField("dy") }
            return .mouseMove(seq: seq, dx: dx, dy: dy)

        case "mouse.click":
            guard let buttonValue = raw.button else { throw MessageValidationError.missingField("button") }
            guard let button = MouseButton(rawValue: buttonValue) else {
                throw MessageValidationError.unsupportedButton(buttonValue)
            }
            return .mouseClick(seq: seq, button: button)

        case "mouse.scroll":
            guard let dx = raw.dx else { throw MessageValidationError.missingField("dx") }
            guard let dy = raw.dy else { throw MessageValidationError.missingField("dy") }
            return .mouseScroll(seq: seq, dx: dx, dy: dy)

        case "keyboard.text":
            guard let text = raw.text else { throw MessageValidationError.missingField("text") }
            return .keyboardText(seq: seq, text: text)

        case "keyboard.key":
            guard let keyValue = raw.key else { throw MessageValidationError.missingField("key") }
            guard let key = KeyboardKey(rawValue: keyValue) else {
                throw MessageValidationError.unsupportedKey(keyValue)
            }
            let modifiers = try (raw.modifiers ?? []).map { value in
                guard let modifier = KeyboardModifier(rawValue: value) else {
                    throw MessageValidationError.unsupportedModifier(value)
                }
                return modifier
            }
            return .keyboardKey(seq: seq, key: key, modifiers: modifiers)

        default:
            throw MessageValidationError.unsupportedType(raw.type)
        }
    }
}

private struct RawMessage: Decodable {
    let seq: Int?
    let type: String
    let dx: Double?
    let dy: Double?
    let button: String?
    let text: String?
    let key: String?
    let modifiers: [String]?
}

File: windows/Protocol/MessageValidator.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;
using AeroPointAgent.Input;

namespace AeroPointAgent.Protocol
{
    public sealed class MessageValidator
    {
        private readonly HashSet<int> _seenSequences = new();

        public MessageValidator() {}

        public AeroPointMessage Validate(string json)
        {
            RawMessage raw;
            try
            {
                raw = JsonSerializer.Deserialize<RawMessage>(json) ?? throw new MessageValidationError("invalid_json", "Failed to deserialize JSON");
            }
            catch (Exception)
            {
                throw new MessageValidationError("invalid_json", "Invalid JSON format");
            }

            if (raw.Type == null)
            {
                throw new MessageValidationError("missing_field", "Missing 'type' field");
            }

            if (!raw.Seq.HasValue)
            {
                throw new MessageValidationError("missing_field", "Missing 'seq' field");
            }

            int seq = raw.Seq.Value;
            lock (_seenSequences)
            {
                if (!_seenSequences.Add(seq))
                {
                    throw new MessageValidationError("duplicate_sequence", $"Duplicate sequence number: {seq}");
                }
            }

            switch (raw.Type)
            {
                case "mouse.move":
                    if (!raw.Dx.HasValue) throw new MessageValidationError("missing_field", "Missing 'dx' field");
                    if (!raw.Dy.HasValue) throw new MessageValidationError("missing_field", "Missing 'dy' field");
                    return new MouseMoveMessage(seq, raw.Dx.Value, raw.Dy.Value);

                case "mouse.click":
                    if (raw.Button == null) throw new MessageValidationError("missing_field", "Missing 'button' field");
                    if (!Enum.TryParse<MouseButton>(raw.Button, true, out var button))
                    {
                        throw new MessageValidationError("unsupported_button", $"Unsupported button: {raw.Button}");
                    }
                    return new MouseClickMessage(seq, button);

                case "mouse.scroll":
                    if (!raw.Dx.HasValue) throw new MessageValidationError("missing_field", "Missing 'dx' field");
                    if (!raw.Dy.HasValue) throw new MessageValidationError("missing_field", "Missing 'dy' field");
                    return new MouseScrollMessage(seq, raw.Dx.Value, raw.Dy.Value);

                case "keyboard.text":
                    if (raw.Text == null) throw new MessageValidationError("missing_field", "Missing 'text' field");
                    return new KeyboardTextMessage(seq, raw.Text);

                case "keyboard.key":
                    if (raw.Key == null) throw new MessageValidationError("missing_field", "Missing 'key' field");
                    if (!Enum.TryParse<KeyboardKey>(raw.Key, true, out var key))
                    {
                        throw new MessageValidationError("unsupported_key", $"Unsupported key: {raw.Key}");
                    }

                    var modifiers = new List<KeyboardModifier>();
                    if (raw.Modifiers != null)
                    {
                        foreach (var modStr in raw.Modifiers)
                        {
                            if (!Enum.TryParse<KeyboardModifier>(modStr, true, out var modifier))
                            {
                                throw new MessageValidationError("unsupported_modifier", $"Unsupported modifier: {modStr}");
                            }
                            modifiers.Add(modifier);
                        }
                    }
                    return new KeyboardKeyMessage(seq, key, modifiers);

                default:
                    throw new MessageValidationError("unsupported_type", $"Unsupported message type: {raw.Type}");
            }
        }

        private class RawMessage
        {
            [JsonPropertyName("seq")]
            public int? Seq { get; set; }

            [JsonPropertyName("type")]
            public string? Type { get; set; }

            [JsonPropertyName("dx")]
            public double? Dx { get; set; }

            [JsonPropertyName("dy")]
            public double? Dy { get; set; }

            [JsonPropertyName("button")]
            public string? Button { get; set; }

            [JsonPropertyName("text")]
            public string? Text { get; set; }

            [JsonPropertyName("key")]
            public string? Key { get; set; }

            [JsonPropertyName("modifiers")]
            public List<string>? Modifiers { get; set; }
        }
    }
}
